[PEP 558 - WIP] bpo-30744: Trace hooks no longer reset closure state

ncoghlan opened the pull request:
(Declined as per the PEP deferral notice in https://github.com/python/peps/pull/3050/files - this reference implementation was useful at the time, but after the frame management changes in Python 3.11 any implementation of the updated semantics in PEP 558 or PEP 667 would be best off starting from scratch. The test cases from this branch might still be useful, but those can always be copied out to a new branch easily enough)

----
Previously, trace hooks running on a nested function
could incorrectly reset the state of a closure cell.

This avoids that by changing the semantics of the
f_locals namespace on function frames to use a
write-through proxy, such that changes are made
immediately rather than being written back some
arbitrary time later.

[PEP 558](https://www.python.org/dev/peps/pep-0558/) and [bpo-17960](https://bugs.python.org/issue17960) cover additional changes
to locals() and frame.f_locals semantics that are
part of this update.

Remaining TODO items before this PR could be considered complete:

- [ ] Resolve #27525 and resync this PR with the main branch
- [ ] Update PEP 558 to cover the performance improvements made to the proxy implementation (lazy initial cache refresh, share a single fast refs mapping between all proxy instances for a given frame)
- [x] Merge in Python 3.11 frame implementation changes
- [x] Fix fast locals proxy implementation to account for Python 3.11 frame implementation changes
- [x] Fix fast locals proxy implementation to follow the design changes in https://github.com/python/peps/pull/1787
- [x] Stop implicitly updating the local variable snapshot when running Python tracing functions
- [ ] Python API documentation updates for `locals()`, `vars()`, and potentially others (check all `locals` mentions)
- [ ] docstring updates for `locals()` (and potentially `vars()`)
- [ ] `PyEval_*` C API documentation updates
- [ ] `PyFrame_*` C API documentation updates
- [x] ~Fix C API header layout to follow modern conventions~ (#18052)
- [x] When rearranging headers, fix the clang warning/error about `PyFrameObject` redefinition
- [ ] Add refcount adjustment info for new C APIs
- [x] Migrate `eval()` default locals namespace to `PyLocals_Get()`
- [x] Migrate `exec()` default locals namespace to `PyLocals_Get()`
- [ ] Report an error if `IMPORT_STAR` opcode is encountered in a `CO_OPTIMIZED` frame
- [x] Add an explicit test for the new `PyFrame_LocalsToFast()` runtime error
- [x] Add explicit C API tests for `PyLocals_Get()` and `PyEval_GetLocals()` at different scopes
- [x] Implement and test the other new PyLocals and PyFrame functions
- [x] Implement and test the following methods on the fast locals proxy:
    - [x] `setdefault()`
    - [x] `popitem()`
    - [x] `clear()`
    - [x] `update()` (theoretically already implemented, but needs explicit tests)
- [ ] Add explicit tests for the behaviour of proxies that reference a cleared frame

Desirable code structure improvements (current plan for these is to put any duplicated into a sharable location in this PR, but file separate maintainability issues to migrate to using the shared code in the original locations, to avoid the diff size impact on this PR):

- [ ] Deduplicate the mutable mapping code copied from `odictobject.c`
- [x] ~Move the `DictProxy` code out of `descrobject.c`~ (obsolete task, as frame locals proxy is now independent of mapping proxy)

<!-- issue-number: [bpo-30744](https://bugs.python.org/issue30744) -->
https://bugs.python.org/issue30744
<!-- /issue-number -->

ncoghlan committed:
bpo-30744: Trace hooks no longer reset closure state

Previously, trace hooks running on a nested function
could incorrectly reset the state of a closure cell.

This avoids that by slightly changing the semantics
of the locals namespace in functions, generators,
and coroutines, such that closure references are
represented in the locals namespace by their
actual cell objects while a trace hook implemented
in Python is running.

Depends on PEP 558 and bpo-17960 to cover the
underlying change to frame.f_locals semantics.

ncoghlan committed:
Disable the current broken writeback logic

ncoghlan committed:
Failing test case for writeback functionality

ncoghlan committed:
Initial skeleton for a write-through proxy

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Finish adding the write-through proxy

ZackerySpytz commented on Objects/frameobject.c:
```diff
@@ -1091,9 +1105,20 @@ fastlocalsproxy_write_to_frame(fastlocalsproxyobject *flp, PyObject *key, PyObje
+                result = -1;
+            } else if (offset > max_offset) {
+                PyErr_Format(PyExc_SystemError,
+                             "Fast locals ref (%d) exceeds array bound (%d)",
```
`%zd` needs to be used with `Py_ssize_t` (these errors were fixed recently in d53fe5f407ff4b529628b01a1bcbf21a6aad5c3a / https://bugs.python.org/issue36254).

ncoghlan replied:
Fixed

auvipy reviewed (changes requested):
rebase needed and build failures needed addressing.

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Add test case for the PEP 558 locals() behaviour

ncoghlan committed:
Actually implement most of the PEP and fix the tests

ncoghlan committed:
Fix segfault on cleared frames

ncoghlan committed:
Use correct printf formatting code

ncoghlan committed:
Initial skeleton for other mutable mapping methods

ncoghlan committed:
Break ref cycle when frame finishes executing

ncoghlan committed:
Remove implicit frame locals update

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Avoid double DECREF on error

ncoghlan committed:
Attempt to make dealloc more robust under gc

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

Update implementation to account for the replaceable frame
evaluation function pointer.

ncoghlan committed:
Fix post-merge compilation errors

ncoghlan committed:
Implement flp.pop()

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Refactor ref map creation

ncoghlan committed:
Correctly manage fast local refcounts

ncoghlan committed:
Allow closure updates after frame termination

ncoghlan committed:
Refactor frame post-eval cleanup

ncoghlan committed:
Use full word in API name

ncoghlan committed:
Update to match latest PEP draft

ncoghlan committed:
Update test_scope for snapshot semantics

ncoghlan committed:
Update test_sys_settrace for snapshot semantics

ncoghlan committed:
Fix pop/delete locals proxy bug

ncoghlan committed:
Update argument clinic output

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Migrate to revised public API design

ncoghlan committed:
Rename PyFrame_LocalsIsSnapshot to PyFrame_GetLocalsReturnsSnapshot

ncoghlan committed:
Mark fast locals proxy as an internal type

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Update draft C API to match latest PEP text

ncoghlan committed:
Migrate exec() and eval() to PyLocals_Get()

ncoghlan committed:
Merge remote-tracking branch 'origin/master' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Avoid circular reference between locals proxy and frame

Mark Shannon pointed out that if every access to the Python
level frame f_locals attributes creates a new write-through
proxy instance, there's no need to actually store the proxy
itself on the frame.

Instead, we can continue to store only the shared mapping
returned by PyEval_GetLocals(), and avoid creating a
circular reference that we then need to break when frame
evaluation ends.

ncoghlan committed:
Add back implicit view refresh in Python trace hook

ncoghlan committed:
Attempt to tidy up Mac OS X compile warnings/errors

ncoghlan committed:
Merge remote-tracking branch 'origin/main' into bpo-30744-make-locals-closure-safe

* Integrates frame internal structure updates from 3.11
* Causes regressions in the PEP 558 implementation (the fast locals proxy
  doesn't get built correctly at the moment)

ncoghlan committed:
Fix Argument Clinic checksum

ncoghlan committed:
Fix stable ABI minimum version

ncoghlan committed:
Bring implementation into line with latest PEP version

ncoghlan committed:
Register new stable ABI additions

ncoghlan committed:
Add FLP str(), fix various value lookup issues

ncoghlan committed:
Uninitialised fields will get you every time

ncoghlan committed:
Fix refcounting, bdb segfault, pdb functionality

ncoghlan committed:
Delegate more operations to the full dynamic snapshot

ncoghlan committed:
Add TODO item for false positives in containment checks

ncoghlan committed:
Only ensure frame snapshot is up to date in O(n) proxy operations

ncoghlan committed:
Keep locals snapshot up to date when reading/writing individual keys

ncoghlan committed:
Avoid false positives in FLP contains method

ncoghlan committed:
Merge remote-tracking branch 'origin/main' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Finish public C API, start dict API tests

ncoghlan committed:
Merge remote-tracking branch 'origin/main' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
Remove debugging print statement

ncoghlan committed:
Regenerated stable ABI files

ncoghlan committed:
Rename _PyLocals_Kind APIs to avoid potential confusion

ncoghlan committed:
Merge remote-tracking branch 'origin/main' into bpo-30744-make-locals-closure-safe

ncoghlan committed:
PyLocals_GetReturnsCopy -> PyLocals_GetKind()

ncoghlan committed:
Share fast_refs mapping between proxy objects

ncoghlan committed:
Remove debugging print

ncoghlan committed:
Defer value cache refresh until needed, start fleshing out dict API tests

ncoghlan committed:
Add dict union operations to proxy

ncoghlan committed:
Implement and test locals proxy clear() method

ncoghlan committed:
Remove pointless print() call

ncoghlan committed:
Implement proxy pop() tests

ncoghlan committed:
Implement and test proxy popitem()

ncoghlan committed:
Test popitem with cells and extra variables

ncoghlan committed:
Implement and test setdefault()

ncoghlan committed:
Implement and test proxy __sizeof__()

ncoghlan committed:
Add C API test for the LocalsToFast exception

ncoghlan committed:
Force enum size

ncoghlan committed:
Clarify code comment

ncoghlan committed:
Keep track of defined names even on cleared frames
